{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "709997af-7bdb-4c87-9251-474937ab4c7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import egglib\n",
    "import glob\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c819942f-63da-413f-b5de-3d88a95c17de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define operator to calculate segregating sites\n",
    "\n",
    "cs = egglib.stats.ComputeStats()\n",
    "cs.add_stats('S')\n",
    "alphabet = egglib.Alphabet('char', ['a', 't', 'c', 'g', '-', 'N'], [], case_insensitive = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "375c8d0d-d715-4d96-b91f-6635127dde0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the number of segregating sites for all opa alignments\n",
    "filenames = glob.glob('../../data/kit14/opa_sequences_aligned/*.aln')\n",
    "\n",
    "filename = '../../data/kit14/opa_sequences_aligned/DGI_65_opa_3.aln'\n",
    "opa_names = []\n",
    "num_different = []\n",
    "for filename in filenames:\n",
    "    opa_name = filename.split('/')[-1][:-4]\n",
    "    aln1 = egglib.io.from_fasta(filename, labels=True, alphabet = alphabet)#egglib.alphabets.DNA)\n",
    "    struct = egglib.struct_from_labels(aln1, lvl_pop=0, lvl_indiv=1)\n",
    "    stats = cs.process_align(aln1, max_missing = 1)\n",
    "    \n",
    "    opa_names.append(opa_name)\n",
    "    num_different.append(stats['S'])\n",
    "    \n",
    "df = pd.DataFrame({'opa_name':opa_names, 'segregating_sites':num_different})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "3f00eeeb-4da2-4cc4-a6bd-4f2e22f3b3bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>opa_name</th>\n",
       "      <th>segregating_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DGI_65_opa_3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DGI_65_opa_8</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EEE029_opa_7</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>DGI_65_opa_10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>EEE029_opa_9</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         opa_name  segregating_sites\n",
       "3    DGI_65_opa_3                  5\n",
       "4    DGI_65_opa_8                  5\n",
       "16   EEE029_opa_7                  5\n",
       "35  DGI_65_opa_10                  5\n",
       "42   EEE029_opa_9                  5"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# These are the opas with non-zero segregating sites\n",
    "df[df['segregating_sites']>0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e17e2c3-84a1-4887-aa64-bbd18f9fd16a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# After inspecting the alignments manually, these are all changes in the repeat regions \n",
    "# in the assemblies where the parent was polished by short reads from the evolved strain with polypolish AND pypolca.\n",
    "# Polishing with polypolish alone and short reads from the evolved strain did not lead to these changes.\n",
    "# Additionally, the long read only assemblies had the same sequence as the short-read polished assemblies."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-egglib]",
   "language": "python",
   "name": "conda-env-.conda-egglib-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
