This directory contains the Snakemake pipeline and downstream Jupyter notebooks to run the main analysis of *opa* diversity, phase, and evolution for the paper. It also contains the figures for the manuscript.

The Snakemake pipeline is compatible with Snakemake v8+ and conda v24.7.1+. 

## Installing snakemake
1. Make a new environment: `conda create --name snakemake8`
2. [Install snakemake](https://snakemake.readthedocs.io/en/stable/getting_started/installation.html): `mamba create -c conda-forge -c bioconda -n snakemake snakemake=8.25.5` (can alternatively use conda)
3. [Install `executor-plugin-cluster-generic`](https://snakemake.github.io/snakemake-plugin-catalog/plugins/executor/cluster-generic.html): `mamba install snakemake-executor-plugin-cluster-generic`

## To run
1. Deposit complete genomes into the directory `input_data/complete_genome_assemblies/`.
2. Submit job using `sbatch start_snakemake.sh` from the main directory.

## Subdirectory structure

### workflow/
This subdirectory contains the code organized into:
* `rules/`: contains the Snakemake rules
* `scripts/`: contains the scripts that the rules depend on
* `notebooks/`: contains the Jupyter notebooks for downstream analysis using outputs from the Snakemake pipeline (mostly making figures) 
* `envs/`: contains the conda environment files used in the Snakemake pipeline
* `Snakefile`: the Snakefile that incorporates the different snakemake rule files and looks for the appropriate overall outputs

### input_data/
This subdirectory is where the pipeline looks for input genomes and where the genome metadata are located. The input genomes are not included in the GitHub repository due to file size limits, but can be found at (insert NCBI Bioproject). The directory is organized as follows:
* `complete_genome_assemblies/`: contains complete genome assemblies (either outputs from Autocycler assemblies of Oxford Nanopore sequences or publicly downloaded complete genomes)
* `complete_genome_pseudogenomes/`: contains pseudogenomes (reference-mapped draft genome assemblies) corresponding to the complete genome assemblies. These were generated by first simulating short reads (see `../draft_genome_assembly/simulate_short_reads`) and then mapping the short reads ot a reference genome (see `../draft_genome_assembly/gc_genomics/`)
* `representative_genome_pseudogenomes/`: containts pseudogenomes (reference-mapped draft genome assemblies) corresponding to representative genomes from the *N. gonorrhoeae* species.
* `whole_genome_metadata/`: contains metadata about complete and representative genomes, including isolate sampling date, location, publication reference, sequencer and assembly method (for complete genomes).

### figures/
Contains figures in the manuscript (plus some bonus figures that didn't make it into the manuscript). The manuscript figures are mostly in the `illustrator/` subdirectory (`fig2.png`, `fig3.png`, `fig5.png`, `fig6_rate_of_evo.png`, `fig7.png`). Figure 1 is at `itol/representative_and_complete_genomes.png` and Figure 4 is at `frame/frame_on_by_strain_compare_null_dist.png`.

### resources/
Contains various reference sequences used by different parts of the Snakemake pipeline. 

### config/
This subdirectory contains the config files for running Snakemake.

## Manual scripts
The following parts of the analysis must be run manually:
- All Jupyter notebooks to analyze the outputs from the Snakemake pipeline (`workflow/notebooks/`)
- iTol tree must be generated in [iTol](https://itol.embl.de/) (`results/itol/complete_genome_pseudogenomes.final_tree.itol_order.tre`)
- BEAST analysis to get time-scaled phylogeny of subtree (Generate the file `results/beast/complete_genomes_recomb_masked.xml` in BEAUTI and then run `workflow/scripts/beast.sh`)
