{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "beadf9fa-01f1-465f-8531-7e2405b65b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "89a275a2-1890-4b4a-abf7-39f8bd9c83c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_22108/1262252152.py:2: DtypeWarning: Columns (19,39,187) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  roary = pd.read_csv('../../results/roary/gene_presence_absence.csv')\n"
     ]
    }
   ],
   "source": [
    "# Load roary and metadata\n",
    "roary = pd.read_csv('../../results/roary/gene_presence_absence.csv')\n",
    "metadata = pd.read_csv('../../results/roary/opa_with_prokka.csv', index_col = 0)\n",
    "opa_metadata = pd.read_csv('../../results/opa_metadata_locus.csv', index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "96c9b043-820c-4cde-81ec-aa995524c17e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['group_10' 'group_1130' 'group_12' 'group_13' 'group_16' 'group_17'\n",
      " 'group_2' 'group_20' 'group_21' 'group_22' 'group_24' 'group_25'\n",
      " 'group_3' 'group_31' 'group_33' 'group_34' 'group_36' 'group_4'\n",
      " 'group_4940' 'group_5' 'group_7' 'group_9' 'piiC_1' 'piiC_10' 'piiC_11'\n",
      " 'piiC_2' 'piiC_3' 'piiC_4' 'piiC_5' 'piiC_6' 'piiC_7' 'piiC_8' 'piiC_9']\n"
     ]
    }
   ],
   "source": [
    "# Figure out which roary clusters the identified opa genes are from\n",
    "roary_clusters = []\n",
    "for i, row in metadata.iterrows():\n",
    "    if len(roary[roary.eq(row['locus_tag']).any(axis=1)])==1:\n",
    "        roary_clusters.append(roary[roary.eq(row['locus_tag']).any(axis=1)]['Gene'].values[0])\n",
    "        \n",
    "unique_roary_clusters = np.unique(roary_clusters)\n",
    "print(unique_roary_clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2a9c9801-b6a2-4208-8ae8-8fe1442604b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes in the roary clusters corresponding to identified opa genes = 2433\n"
     ]
    }
   ],
   "source": [
    "# Get the information about these roary clusters\n",
    "roary_opa = pd.DataFrame(unique_roary_clusters, columns = ['Gene']).merge(roary)\n",
    "print('Number of genes in the roary clusters corresponding to identified opa genes = ' + str(np.sum(roary_opa['No. sequences'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "16f65258-a1b2-4145-a758-06ce94f70c26",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat the results so that each row contains the locus tag (above, each row contained the entire gene presence absence table)\n",
    "general_columns = ['Gene', \n",
    "                   'Non-unique Gene name',\n",
    "                   'Annotation', \n",
    "                   'No. isolates',\n",
    "                   'No. sequences',\n",
    "                   'Avg sequences per isolate',\n",
    "                   'Genome Fragment',\n",
    "                   'Order within Fragment',\n",
    "                   'Accessory Fragment',\n",
    "                   'Accessory Order with Fragment',\n",
    "                   'QC',\n",
    "                   'Min group size nuc',\n",
    "                   'Max group size nuc',\n",
    "                   'Avg group size nuc']\n",
    "strains = np.sort(list(set(roary_opa.columns) - set(general_columns)))\n",
    "\n",
    "roary_opa_by_gene = pd.DataFrame()\n",
    "for strain in strains:\n",
    "    roary_opa_by_gene_strain = roary_opa[['Gene', 'Non-unique Gene name', 'Annotation', 'No. isolates', 'No. sequences', strain]].dropna(subset = [strain])\n",
    "    roary_opa_by_gene_strain.rename({'Gene':'roary', 'Non-unique Gene name':'prokka_gene_name', strain:'locus_tag'}, axis = 'columns', inplace = True)\n",
    "    roary_opa_by_gene = pd.concat([roary_opa_by_gene,roary_opa_by_gene_strain])\n",
    "\n",
    "roary_opa_by_gene.reset_index(inplace = True, drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "68ccc656-e039-4a4b-86bf-ba4731f5aae2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_22108/492599154.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  roary_only['locus_tag'] = roary_only['locus_tag'].str.split(expand = True)[0]\n"
     ]
    }
   ],
   "source": [
    "# Figure out which ones were already found by our script\n",
    "merged = roary_opa_by_gene.merge(metadata, on = 'locus_tag', how = 'outer', indicator = True)\n",
    "\n",
    "metadata_only = merged[merged['_merge'] == 'right_only']\n",
    "metadata_only.reset_index(drop = True, inplace = True)\n",
    "\n",
    "roary_only = merged[merged['_merge'] == 'left_only']\n",
    "roary_only.reset_index(drop = True, inplace = True)\n",
    "\n",
    "# If there were two locus tags associated with an entry (may have happened if the ORF was split), only keep the first locus tag\n",
    "roary_only['locus_tag'] = roary_only['locus_tag'].str.split(expand = True)[0]\n",
    "\n",
    "# Get the locations of the genes from Prokka for the genes that were only identified by roary but not our algorithm\n",
    "prokka_path = '../../results/annotations/'\n",
    "\n",
    "filenames = glob.glob(prokka_path + '*.csv')\n",
    "filenames = np.sort(filenames)\n",
    "prokka_all = pd.DataFrame()\n",
    "for filename in filenames:\n",
    "    strain = os.path.basename(filename)[:-4]\n",
    "    prokka = pd.read_csv(filename, index_col = 0)\n",
    "    prokka['strain'] = strain\n",
    "    prokka_all = pd.concat([prokka_all, prokka])\n",
    "prokka_all.reset_index(drop = True, inplace = True)\n",
    "\n",
    "roary_only = roary_only.drop(['chromosome_metadata', 'strand', 'start_cr', 'stop_cr', 'start_term', 'stop_term', 'id', 'start_metadata', 'stop', 'n_terminus', 'in_frame', 'chromosome_prokka', 'source', '_merge', 'strain', 'gene', 'start_prokka', 'end', 'lcb_num', 'ref_lcb_start',\n",
    "       'ref_lcb_stop', 'ref_lcb_strand', 'query_lcb_start', 'query_lcb_stop',\n",
    "       'query_lcb_strand', 'length_ratio', 'start_reordered',\n",
    "       'start_reordered_flipped', 'strand_flipped', 'locus'], axis = 'columns')\n",
    "roary_only = roary_only.merge(prokka_all, on = 'locus_tag', how = 'left')\n",
    "\n",
    "# Check if the roary only hits were actually just ORFs split into 2 ORFs - use very strict cutoff since I would rather have more false positives than false negatives\n",
    "roary_only.sort_values('start', inplace = True, ignore_index = True)\n",
    "opa_metadata.sort_values('start', inplace = True, ignore_index = True)\n",
    "\n",
    "roary_only = roary_only[pd.merge_asof(roary_only, opa_metadata, on = 'start', by=['strain','strand'], direction = 'nearest', tolerance = 700)['id'].isnull()]\n",
    "roary_only.to_csv('../../results/roary/roary_only.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "da83002b-944a-4f47-bcea-89c89086f1e0",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get the sequences\n",
    "path = '../../results/assemblies_shifted/'\n",
    "roary_only_records = []\n",
    "i = 0\n",
    "for i, row in roary_only.iterrows():\n",
    "    strain = row['strain']\n",
    "    start = int(row['start'])\n",
    "    stop = int(row['end'])\n",
    "    strand = row['strand']\n",
    "    \n",
    "    record = list(SeqIO.parse(path + strain + '.fa', \"fasta\"))[0]\n",
    "    \n",
    "    if strand == 1: \n",
    "        roary_only_records.append(SeqRecord(record.seq[start-100:start+900], id = strain + '_' + str(start) + '_' + str(stop)))\n",
    "    elif strand == -1:\n",
    "        roary_only_records.append(SeqRecord(record.seq[stop-900:stop+100].reverse_complement(), id = strain + '_' + str(start) + '_' + str(stop)))\n",
    "        \n",
    "    # Add FA1090 sequences, with extra sequence before and after to test for deletions\n",
    "for i, row in metadata[metadata['strain']=='FA1090'].iterrows():\n",
    "    strain = row['strain']\n",
    "    start = int(row['start_metadata'])\n",
    "    stop = int(row['stop'])\n",
    "    strand = row['strand']\n",
    "    \n",
    "    record = list(SeqIO.parse(path + 'FA1090.fa', \"fasta\"))[0]\n",
    "    if strand == 1: \n",
    "        roary_only_records.append(SeqRecord(record.seq[start:stop], id = row['id']))\n",
    "    elif strand == -1:\n",
    "        roary_only_records.append(SeqRecord(record.seq[start:stop].reverse_complement(), id = row['id']))\n",
    "        \n",
    "SeqIO.write(roary_only_records, '../../results/roary/opa_roary_only_with_FA1090.fa', 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "bb2bc934-670a-4795-9738-ef08f60a2ea2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Number of opa genes found by our script that were not found by Roary: 7'"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'Number of opa genes found by our script that were not found by Roary: ' + str(len(metadata_only))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "179a0d6a-fbf2-4035-a1f3-030f0f0b052e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      10794_opa_10\n",
       "1        FQ82_opa_5\n",
       "2       9460_opa_10\n",
       "3       9464_opa_10\n",
       "4     FFF007_opa_11\n",
       "5    GCGS0423_opa_2\n",
       "6     JJJ010_opa_11\n",
       "Name: id, dtype: object"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_only['id']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-biopython]",
   "language": "python",
   "name": "conda-env-.conda-biopython-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
