{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "709997af-7bdb-4c87-9251-474937ab4c7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import egglib\n",
    "import glob\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c819942f-63da-413f-b5de-3d88a95c17de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define operator to calculate segregating sites\n",
    "\n",
    "cs = egglib.stats.ComputeStats()\n",
    "cs.add_stats('S')\n",
    "alphabet = egglib.Alphabet('char', ['a', 't', 'c', 'g', '-', 'N'], [], case_insensitive = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "375c8d0d-d715-4d96-b91f-6635127dde0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the number of segregating sites for all opa alignments\n",
    "filenames = glob.glob('../../data/opa_sequences_aligned/*.aln')\n",
    "\n",
    "opa_names = []\n",
    "num_different = []\n",
    "for filename in filenames:\n",
    "    opa_name = filename.split('/')[-1][:-4]\n",
    "    aln1 = egglib.io.from_fasta(filename, labels=True, alphabet = alphabet)#egglib.alphabets.DNA)\n",
    "    struct = egglib.struct_from_labels(aln1, lvl_pop=0, lvl_indiv=1)\n",
    "    stats = cs.process_align(aln1, max_missing = 1)\n",
    "    \n",
    "    opa_names.append(opa_name)\n",
    "    num_different.append(stats['S'])\n",
    "    \n",
    "df = pd.DataFrame({'opa_name':opa_names, 'segregating_sites':num_different})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f00eeeb-4da2-4cc4-a6bd-4f2e22f3b3bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>opa_name</th>\n",
       "      <th>segregating_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>EEE023_opa_4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>FFF007_opa_11</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>GCGS0313_opa_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>GCGS0313_opa_5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           opa_name  segregating_sites\n",
       "19     EEE023_opa_4                  1\n",
       "79    FFF007_opa_11                 24\n",
       "118  GCGS0313_opa_1                  1\n",
       "95   GCGS0313_opa_5                  1"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# These are the opas with non-zero segregating sites\n",
    "df[df['segregating_sites']>0].sort_values('opa_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "227c9461",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if there were any assemblies where an opa was missing.\n",
    "filename = '../../data/opa_locations/UMASS-DGI_65_50x_1.csv'\n",
    "filenames = glob.glob('../../data/opa_locations/*.csv')\n",
    "\n",
    "opa_metadata = pd.DataFrame()\n",
    "for filename in filenames:\n",
    "    basename = os.path.basename(filename)[:-4]\n",
    "    basename_split = basename.split('_')\n",
    "    read_depth = int(basename_split[-2][:-1])\n",
    "    read_set = int(basename_split[-1])\n",
    "    strain = '_'.join(basename_split[:-2])\n",
    "    \n",
    "    df = pd.read_csv(filename, index_col = 0)\n",
    "    df['strain'] = strain\n",
    "    df['depth'] = read_depth\n",
    "    df['set'] = read_set\n",
    "    df['id'] = strain + '_opa_' + df['id'].str.split('_opa_', expand = True)[1]\n",
    "    \n",
    "    opa_metadata = pd.concat([opa_metadata, df], ignore_index = True)\n",
    "    \n",
    "opa_metadata.sort_values(['strain', 'depth', 'set'], inplace = True, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "8a85911e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>start_cr</th>\n",
       "      <th>stop_cr</th>\n",
       "      <th>start_term</th>\n",
       "      <th>stop_term</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>n_terminus</th>\n",
       "      <th>id</th>\n",
       "      <th>in_frame</th>\n",
       "      <th>strain</th>\n",
       "      <th>depth</th>\n",
       "      <th>set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>69624.0</td>\n",
       "      <td>69664.0</td>\n",
       "      <td>68898.0</td>\n",
       "      <td>68916.0</td>\n",
       "      <td>68899.0</td>\n",
       "      <td>69689.0</td>\n",
       "      <td>69610.0</td>\n",
       "      <td>CCC020_opa_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CCC020</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>75119.0</td>\n",
       "      <td>75189.0</td>\n",
       "      <td>74396.0</td>\n",
       "      <td>74414.0</td>\n",
       "      <td>74397.0</td>\n",
       "      <td>75214.0</td>\n",
       "      <td>75105.0</td>\n",
       "      <td>CCC020_opa_2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CCC020</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>983486.0</td>\n",
       "      <td>983496.0</td>\n",
       "      <td>984225.0</td>\n",
       "      <td>984243.0</td>\n",
       "      <td>983463.0</td>\n",
       "      <td>984242.0</td>\n",
       "      <td>983510.0</td>\n",
       "      <td>CCC020_opa_3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CCC020</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1061455.0</td>\n",
       "      <td>1061525.0</td>\n",
       "      <td>1060705.0</td>\n",
       "      <td>1060723.0</td>\n",
       "      <td>1060706.0</td>\n",
       "      <td>1061548.0</td>\n",
       "      <td>1061441.0</td>\n",
       "      <td>CCC020_opa_4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CCC020</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1102178.0</td>\n",
       "      <td>1102198.0</td>\n",
       "      <td>1102924.0</td>\n",
       "      <td>1102942.0</td>\n",
       "      <td>1102153.0</td>\n",
       "      <td>1102941.0</td>\n",
       "      <td>1102212.0</td>\n",
       "      <td>CCC020_opa_5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CCC020</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4755</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>541991.0</td>\n",
       "      <td>542081.0</td>\n",
       "      <td>542792.0</td>\n",
       "      <td>542810.0</td>\n",
       "      <td>541966.0</td>\n",
       "      <td>542809.0</td>\n",
       "      <td>542095.0</td>\n",
       "      <td>UMASS-DGI_65_opa_6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>UMASS-DGI_65</td>\n",
       "      <td>125</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4756</th>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>620319.0</td>\n",
       "      <td>620364.0</td>\n",
       "      <td>619575.0</td>\n",
       "      <td>619593.0</td>\n",
       "      <td>619576.0</td>\n",
       "      <td>620390.0</td>\n",
       "      <td>620305.0</td>\n",
       "      <td>UMASS-DGI_65_opa_7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>UMASS-DGI_65</td>\n",
       "      <td>125</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4757</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1533651.0</td>\n",
       "      <td>1533734.0</td>\n",
       "      <td>1534442.0</td>\n",
       "      <td>1534460.0</td>\n",
       "      <td>1533626.0</td>\n",
       "      <td>1534459.0</td>\n",
       "      <td>1533748.0</td>\n",
       "      <td>UMASS-DGI_65_opa_8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>UMASS-DGI_65</td>\n",
       "      <td>125</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4758</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1583183.0</td>\n",
       "      <td>1583198.0</td>\n",
       "      <td>1583924.0</td>\n",
       "      <td>1583942.0</td>\n",
       "      <td>1583160.0</td>\n",
       "      <td>1583941.0</td>\n",
       "      <td>1583212.0</td>\n",
       "      <td>UMASS-DGI_65_opa_9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>UMASS-DGI_65</td>\n",
       "      <td>125</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4759</th>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1905336.0</td>\n",
       "      <td>1905396.0</td>\n",
       "      <td>1904580.0</td>\n",
       "      <td>1904598.0</td>\n",
       "      <td>1904581.0</td>\n",
       "      <td>1905420.0</td>\n",
       "      <td>1905322.0</td>\n",
       "      <td>UMASS-DGI_65_opa_10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>UMASS-DGI_65</td>\n",
       "      <td>125</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4760 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome  strand   start_cr    stop_cr  start_term  stop_term  \\\n",
       "0              1      -1    69624.0    69664.0     68898.0    68916.0   \n",
       "1              1      -1    75119.0    75189.0     74396.0    74414.0   \n",
       "2              1       1   983486.0   983496.0    984225.0   984243.0   \n",
       "3              1      -1  1061455.0  1061525.0   1060705.0  1060723.0   \n",
       "4              1       1  1102178.0  1102198.0   1102924.0  1102942.0   \n",
       "...          ...     ...        ...        ...         ...        ...   \n",
       "4755           1       1   541991.0   542081.0    542792.0   542810.0   \n",
       "4756           1      -1   620319.0   620364.0    619575.0   619593.0   \n",
       "4757           1       1  1533651.0  1533734.0   1534442.0  1534460.0   \n",
       "4758           1       1  1583183.0  1583198.0   1583924.0  1583942.0   \n",
       "4759           1      -1  1905336.0  1905396.0   1904580.0  1904598.0   \n",
       "\n",
       "          start       stop  n_terminus                   id  in_frame  \\\n",
       "0       68899.0    69689.0     69610.0         CCC020_opa_1       0.0   \n",
       "1       74397.0    75214.0     75105.0         CCC020_opa_2       0.0   \n",
       "2      983463.0   984242.0    983510.0         CCC020_opa_3       0.0   \n",
       "3     1060706.0  1061548.0   1061441.0         CCC020_opa_4       0.0   \n",
       "4     1102153.0  1102941.0   1102212.0         CCC020_opa_5       0.0   \n",
       "...         ...        ...         ...                  ...       ...   \n",
       "4755   541966.0   542809.0    542095.0   UMASS-DGI_65_opa_6       1.0   \n",
       "4756   619576.0   620390.0    620305.0   UMASS-DGI_65_opa_7       0.0   \n",
       "4757  1533626.0  1534459.0   1533748.0   UMASS-DGI_65_opa_8       0.0   \n",
       "4758  1583160.0  1583941.0   1583212.0   UMASS-DGI_65_opa_9       0.0   \n",
       "4759  1904581.0  1905420.0   1905322.0  UMASS-DGI_65_opa_10       0.0   \n",
       "\n",
       "            strain  depth  set  \n",
       "0           CCC020     50    1  \n",
       "1           CCC020     50    1  \n",
       "2           CCC020     50    1  \n",
       "3           CCC020     50    1  \n",
       "4           CCC020     50    1  \n",
       "...            ...    ...  ...  \n",
       "4755  UMASS-DGI_65    125   10  \n",
       "4756  UMASS-DGI_65    125   10  \n",
       "4757  UMASS-DGI_65    125   10  \n",
       "4758  UMASS-DGI_65    125   10  \n",
       "4759  UMASS-DGI_65    125   10  \n",
       "\n",
       "[4760 rows x 14 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opa_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23a07690",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
