{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "0ef663c4-278c-493c-a300-1b54d44464f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import glob\n",
    "import shutil\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "28dd4801-5125-4872-be22-a2092ecb5e88",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5plus1\n",
      "5plus2\n",
      "5plus3\n",
      "5plus4\n",
      "5minus1\n",
      "5minus2\n",
      "5minus3\n",
      "5minus4\n",
      "3minus3\n",
      "3plus5\n",
      "3plus6\n",
      "3plus7\n",
      "3plus8\n",
      "3plus10\n",
      "3plus13\n",
      "3plus15\n",
      "65bh_8\n",
      "65bd_8\n",
      "65bSa_8\n",
      "65bSb_8\n",
      "EEE020\n",
      "copied\n",
      "EEE020d_8\n",
      "EEE020a_8\n",
      "EEE020b_8\n",
      "EEE013\n",
      "GCGS0306\n",
      "GCGS0313\n",
      "UMASS-DGI_5\n",
      "UMASS-DGI_9\n",
      "copied\n",
      "DGI9_IIa\n",
      "DGI9_IIb\n",
      "DGI9_IIc\n",
      "DGI9_IIdeltaNHS\n",
      "HHH042\n",
      "GCGS0423\n",
      "FA6140\n",
      "JJJ020\n",
      "DDD038\n",
      "UMASS-DGI_1\n",
      "GGG031\n",
      "HHH030\n",
      "UMASS-DGI_54\n",
      "FFF017\n",
      "FFF012\n",
      "copied\n",
      "FFF24bV\n",
      "HHH023\n",
      "HHH023aVIII\n",
      "HHH023bVIII\n",
      "HHH023dNHSVIII\n",
      "GCGS1013aII\n",
      "GCGS1013bII\n",
      "GCGS1013cII\n",
      "GCGS1013\n",
      "HHH023HBSSVIII\n",
      "FFF024\n",
      "FFF024dNHSV\n",
      "FFF024HBSSV\n",
      "DDD046\n",
      "UMASS-DGI_26\n",
      "JJJ025\n",
      "GCGS0306\n",
      "GGG022\n",
      "UMASS-DGI_36\n",
      "copied\n",
      "GCGS0769\n",
      "96MB\n",
      "EEE018\n",
      "FFF038\n",
      "DDD031\n",
      "EEE004\n",
      "GCGS0245\n",
      "GCGS0276\n",
      "GCGS0306\n",
      "UMASS-DGI_40\n",
      "FFF007\n",
      "UMASS-DGI_26\n",
      "Ng_188\n",
      "copied\n",
      "25818_3#234\n",
      "copied\n",
      "GCGS0834\n",
      "GCGS0339\n",
      "Ng_118\n",
      "copied\n",
      "GCGS1014\n",
      "GCGS1095\n",
      "FA19\n",
      "GCGS0457\n",
      "MA_parent\n",
      "MA_A\n",
      "MA_D\n",
      "MA_E\n",
      "MA_G\n",
      "MA_H\n",
      "GCGS0864\n",
      "EEE029\n",
      "EEE029a_VI\n",
      "EEE029a_VI_8passages\n",
      "EEE029b_VI\n",
      "EEE029b_VI_8passages\n",
      "EEE029_dNHS_VI\n",
      "UMASS-DGI_65\n",
      "DGI_65a_II\n",
      "DGI_65b_IV\n",
      "DGI_65_dNHS_IV\n",
      "GCGS0431\n",
      "GCGS0431a_IV_8passages\n",
      "GCGS0431a_IV\n",
      "GCGS0431b_IV\n",
      "GCGS0431b_IV_8passages\n",
      "GCGS0431_dNHS_IV\n",
      "CCC010\n",
      "CCC012\n",
      "CCC015\n",
      "EEE021\n",
      "EEE026\n",
      "FFF044\n",
      "JJJ010\n",
      "SRR5827322\n",
      "copied\n",
      "AUNG1291\n",
      "SRR5827236\n",
      "copied\n",
      "CCC014\n",
      "EEE015\n",
      "DDD029\n",
      "CCC016\n",
      "FFF026\n",
      "FA1090\n",
      "UMASS-DGI_46\n",
      "FFF004\n",
      "EEE034\n",
      "AUNG1065\n",
      "DDD001\n",
      "CCC017\n",
      "FFF005\n",
      "EEE019\n",
      "175B\n",
      "175D\n",
      "194parent\n",
      "194E\n",
      "194F\n",
      "EEE030\n",
      "FA19\n",
      "CCC048\n",
      "CCC020\n",
      "CCC028\n",
      "EEE014\n",
      "EEE036\n",
      "FFF047\n",
      "EEE037\n",
      "EEE023\n",
      "AUNG1251\n",
      "GCGS0373\n",
      "CCC013\n",
      "EEE035\n",
      "DDD034\n",
      "EEE016\n",
      "CCC033\n",
      "175parent\n"
     ]
    }
   ],
   "source": [
    "# Create directory where the renamed read files will be saved\n",
    "os.makedirs('/n/holylfs05/LABS/grad_lab/Users/qinqinyu/datasets/opa_manuscript_ncbi_submission/nanopore_sequencing_reads/bams/', exist_ok=True)\n",
    "\n",
    "# Get names of all complete genomes sequenced in the Grad lab used in the study\n",
    "complete_genomes_metadata = pd.read_csv('/n/grad_lab2/Users/qinqinyu/projects/opa_project/opa_diversity_snakemake/input_data/whole_genome_metadata/grad_lab_complete_genomes_metadata.csv', index_col = 0)\n",
    "complete_genomes = complete_genomes_metadata['wgs_id'].values\n",
    "\n",
    "# Loop through folders\n",
    "folders = ['230525', '230621', '230802', '230913', '231031', '240821', '240829_gc_qy_representative_genomes_batch2', '241009_gc_sb_qy_collab_representative_genomes_batch3']\n",
    "dest = '/n/holylfs05/LABS/grad_lab/Users/qinqinyu/datasets/opa_manuscript_ncbi_submission/nanopore_sequencing_reads/bams/'\n",
    "\n",
    "for folder in folders:\n",
    "    # Get the correspondence between sample name and barcode number\n",
    "    samples = pd.read_csv('/n/grad_lab2/Lab/gonococcus/nanopore_sequencing/' + folder + '/samples.txt', sep = '\\t', index_col = False)\n",
    "    \n",
    "    # Loop through the read files with the \"barcode\" names\n",
    "    read_files = np.sort(glob.glob('/n/grad_lab2/Lab/gonococcus/nanopore_sequencing/' + folder + '/demultiplexed_reads/*barcode*.bam'))\n",
    "    for read_file in read_files:\n",
    "        # Get the barcode name from the read file\n",
    "        barcode = read_file.split('SQK-NBD114-24_')[1][:-4]\n",
    "        # Get the corresponding wgs id\n",
    "        wgs_ids = samples['wgs_id'][samples['barcode']==barcode].values\n",
    "        \n",
    "        # Check if the corresponding wgs id wasn't found, then use just the barcode number (instead of 'barcode##')\n",
    "        if len(wgs_ids) == 0:\n",
    "            wgs_ids = samples['wgs_id'][samples['barcode']==int(barcode.split('barcode')[1])].values\n",
    "        \n",
    "        # If there was only one wgs id corresponding to this barcode (what we would expect)\n",
    "        if len(wgs_ids) == 1:\n",
    "            # Get that wgs id\n",
    "            wgs_id = wgs_ids[0]\n",
    "            # Check if the wgs id is in the list of complete genomes that we sequenced (since some of the nanopore data is of lab-derived isolates)\n",
    "            print(wgs_id)\n",
    "            if wgs_id in complete_genomes:\n",
    "                # Rename JJJ025 (unpublished isolate from SSS project) to EEE024 (identical published isolate - what is used in the manuscript)\n",
    "                if wgs_id == 'JJJ025':\n",
    "                    wgs_id = 'EEE024'\n",
    "                # If the read file has not previously been copied over, then copy it over\n",
    "                if not os.path.exists(dest + wgs_id + '.bam'):\n",
    "                    shutil.copy(read_file, dest + wgs_id + '.bam') \n",
    "                    print('copied')\n",
    "        \n",
    "        # If there was more than one wgs id corresponding to this barcode, then throw a message\n",
    "        elif len(wgs_ids)>1: \n",
    "            print('More than one wgs id found for this barcode')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "a6f84fa0-b1dc-49f7-aa46-1199382c25a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'241009_gc_sb_qy_collab_representative_genomes_batch3'"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "6a2294af-2545-48bc-91aa-ad624fd94482",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['barcode wgs_id', 'Strain Name', 'Project'], dtype='object')"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "f55496ab-d73d-4a8b-8505-d7ced623b5cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['wgs_idÂ ', 'barcode'], dtype='object')"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "e11d2e8c-ca07-43fd-8205-eb490170bcc7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>wgs_id</th>\n",
       "      <th>Strain Name</th>\n",
       "      <th>Project</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>175B</td>\n",
       "      <td>Melis GC01</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>175D</td>\n",
       "      <td>Melis GC01</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>194parent</td>\n",
       "      <td>Melis GC01</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>194E</td>\n",
       "      <td>Melis GC01</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "      <td>194F</td>\n",
       "      <td>Melis GC01</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7</td>\n",
       "      <td>EEE030</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>8</td>\n",
       "      <td>FA19</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9</td>\n",
       "      <td>CCC048</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>10</td>\n",
       "      <td>CCC020</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>11</td>\n",
       "      <td>CCC028</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>12</td>\n",
       "      <td>EEE014</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>13</td>\n",
       "      <td>EEE036</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>14</td>\n",
       "      <td>FFF047</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>15</td>\n",
       "      <td>EEE037</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>16</td>\n",
       "      <td>EEE023</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>17</td>\n",
       "      <td>AUNG1251</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>18</td>\n",
       "      <td>GCGS0373</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>19</td>\n",
       "      <td>CCC013</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>20</td>\n",
       "      <td>EEE035</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>21</td>\n",
       "      <td>DDD034</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>22</td>\n",
       "      <td>EEE016</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>23</td>\n",
       "      <td>CCC033</td>\n",
       "      <td>QinQin representative genomes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>24</td>\n",
       "      <td>175parent</td>\n",
       "      <td>Melis GC01</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    barcode     wgs_id                    Strain Name  Project\n",
       "0         2       175B                     Melis GC01      NaN\n",
       "1         3       175D                     Melis GC01      NaN\n",
       "2         4  194parent                     Melis GC01      NaN\n",
       "3         5       194E                     Melis GC01      NaN\n",
       "4         6       194F                     Melis GC01      NaN\n",
       "5         7     EEE030  QinQin representative genomes      NaN\n",
       "6         8       FA19  QinQin representative genomes      NaN\n",
       "7         9     CCC048  QinQin representative genomes      NaN\n",
       "8        10     CCC020  QinQin representative genomes      NaN\n",
       "9        11     CCC028  QinQin representative genomes      NaN\n",
       "10       12     EEE014  QinQin representative genomes      NaN\n",
       "11       13     EEE036  QinQin representative genomes      NaN\n",
       "12       14     FFF047  QinQin representative genomes      NaN\n",
       "13       15     EEE037  QinQin representative genomes      NaN\n",
       "14       16     EEE023  QinQin representative genomes      NaN\n",
       "15       17   AUNG1251  QinQin representative genomes      NaN\n",
       "16       18   GCGS0373  QinQin representative genomes      NaN\n",
       "17       19     CCC013  QinQin representative genomes      NaN\n",
       "18       20     EEE035  QinQin representative genomes      NaN\n",
       "19       21     DDD034  QinQin representative genomes      NaN\n",
       "20       22     EEE016  QinQin representative genomes      NaN\n",
       "21       23     CCC033  QinQin representative genomes      NaN\n",
       "22       24  175parent                     Melis GC01      NaN"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('/n/grad_lab2/Lab/gonococcus/nanopore_sequencing/' + folder + '/samples.txt', sep = '\\t', index_col = False)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8148765-6e0e-4c02-9326-2bccab69e233",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
