{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "98e251b8-24e9-4001-b5fb-982e819e607c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import glob\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "dac4cb05-d8bb-4fa8-82c5-a879037c838d",
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblies_path = \"../../../../opa_diversity_snakemake/input_data/complete_genome_assemblies/grad_lab/\"\n",
    "chromosome_list_path = \"../data/assembly_chromosome_list_files/\"\n",
    "os.makedirs(chromosome_list_path, exist_ok=True)\n",
    "\n",
    "filenames = glob.glob(assemblies_path + '*.fa')\n",
    "for filename in filenames:\n",
    "    # Get isolate name\n",
    "    isolate = filename.split('/')[-1][:-3]\n",
    "    \n",
    "    # Load records from fasta file\n",
    "    records = list(SeqIO.parse(filename, \"fasta\"))\n",
    "    \n",
    "    # Open new file for writing chromosome list file\n",
    "    with open(chromosome_list_path + isolate + \".txt\", \"w\") as file:\n",
    "        \n",
    "        # Loop through records (chromosome and plasmids)\n",
    "        for record in records:\n",
    "            \n",
    "            # Get names and lengths of each entry in fasta file\n",
    "            name = record.id\n",
    "            description = record.description.split(' ')\n",
    "            length = int(description[1].split('=')[1])\n",
    "            \n",
    "            # Check if it's the chromosome\n",
    "            if length>2*10**6:\n",
    "                chromosome_type = 'Chromosome'\n",
    "            else:\n",
    "                chromosome_type = 'Plasmid'\n",
    "\n",
    "            # Write to chromosome file\n",
    "            file.write(name + '\\t' + name + '\\t' + chromosome_type + '\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-biopython]",
   "language": "python",
   "name": "conda-env-.conda-biopython-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
